### PR TITLE
[SYCL] Import/preprocess boost/mp11 at build time to use in SYCL headers.

### PR DESCRIPTION
**Pull request by kbobrovs**
Preprocessing adapts boost/mp11 headers for use in SYCL headers in a way
that does not conflict with potential use of boost in user code.
Particularly, `BOOST_*` macros are replaced with `SYCL_BOOST_*`, APIs
are moved into the top-level `sycl::detail` namespace. For example,
`sycl::detail::boost::mp11::mp_list`.

Signed-off-by: Konstantin S Bobrovsky <konstantin.s.bobrovsky@intel.com>

### CODE REVIEW

**Comment by gmlueck** [comment]
> One thing that needs updating for sure is the sycl::boost namespace. Please suggest variants.

We should not pollute the `::sycl` namespace.  Implementation details go in `::sycl::detail`, so maybe `::sycl::detail::boost`.

**Comment by kbobrovs** [comment]
> > One thing that needs updating for sure is the sycl::boost namespace. Please suggest variants.
> 
> We should not pollute the `::sycl` namespace. Implementation details go in `::sycl::detail`, so maybe `::sycl::detail::boost`.

Right, we should not. @rolandschulz also suggested `::sycl::detail::boost` - I will make the change.

**Review by rolandschulz — DISMISSED** [review]

**Review by pvchupin — COMMENTED** [review]
@alexbatashev, can you also review this please? I'm sure you'll have some good suggestions.

File: sycl/CMakeLists.txt
Comment (kbobrovs) [reply sycl/CMakeLists.txt:129]: Good point. I refactored the code. But I can't really add it as llvm project, as we talked, so I put mp11 modules into sycl/cmake/modules, leaving only `include(<main_mp11_module_name>)` in the sycl CMakeLists.txt.

Comment (kbobrovs) [reply sycl/CMakeLists.txt:129]: also I fixed errors arising with cmake 3.16 used in CI

File: sycl/cmake/modules/PreprocessBoostMp11Headers.cmake
Comment (keryell) [inline sycl/cmake/modules/PreprocessBoostMp11Headers.cmake:106]: EOL

**Review by keryell — COMMENTED** [review]
Nice, we can use Boost in the implementation? :-)
Just curious, why do you need MP11?

**Comment by kbobrovs** [comment]
> Nice, we can use Boost in the implementation? :-)
> Just curious, why do you need MP11?

Sort of. We can only boost::mp11 in our implementation for now :)
This is needed for invoke_simd and property_lists @steffenlarsen works on.

File: sycl/cmake/modules/PreprocessBoostMp11Headers.cmake
Comment (kbobrovs) [reply sycl/cmake/modules/PreprocessBoostMp11Headers.cmake]: thanks, fixed

**Comment by kbobrovs** [comment]
@pvchupin, @keryell, @gmlueck - please review, comments addressed.

@intel/llvm-reviewers-runtime, ping

**Review by pvchupin — DISMISSED** [review]

**Review by keryell — DISMISSED** [review]

**Review by pvchupin — DISMISSED** [review]

**Review by gmlueck — DISMISSED** [review]
The header file path `<sycl/detail/boost/mp11.hpp>` and the namespace `sycl::detail::boost::mp11` both look OK tome.

**Review by v-klochkov — APPROVED** [review]
Looks good to me.
I advice to ask for one more reviewer to look at these CMake change.

**Review by alexbatashev — APPROVED** [review]

File: sycl/cmake/modules/PreprocessBoostMp11Headers.cmake
Comment (bader) [inline sycl/cmake/modules/PreprocessBoostMp11Headers.cmake:5]:
```suggestion
# adds SYCL_ prefix to boost macros names. See more specific comments in the
```

**Review by bader — APPROVED** [review]

**Comment by kbobrovs** [comment]
Fixed comment according to review comments in [7cacacb].
@bader, @alexbatashev, @v-klochkov, @gmlueck, @pvchupin  - please re-approve.

**Review by bader — APPROVED** [review]
> Fixed comment according to review comments in [[7cacacb](https://github.com/intel/llvm/commit/7cacacb3d9863dc82d5806cc7b377ff2c8a7e467)]. @bader, @alexbatashev, @v-klochkov, @gmlueck, @pvchupin - please re-approve.

I think this typo fix won't change the status and we can consider this PR approved by all these folks. Are you still waiting for reviews from @sergey-semenov, @steffenlarsen and @rolandschulz?

**Review by gmlueck — APPROVED** [review]

**Comment by kbobrovs** [comment]
> I think this typo fix won't change the status and we can consider this PR approved by all these folks. Are you still waiting for reviews from @sergey-semenov, @steffenlarsen and @rolandschulz?

No, I think I have enough reviews at this point.

**Review by pvchupin — APPROVED** [review]

**Review by alexbatashev — APPROVED** [review]

**Comment by kbobrovs** [comment]
@steffenlarsen, FYI - you can now use mp11

**Pull request merged**
